### PR TITLE
SparkSQL: Update terminator grammar for `HAVING`, `WHERE`, `GROUP BY`

### LINKED ISSUE #3484: SparkSQL dialect templating/parsing errors found with WHERE and DISTRIBUTE BY
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

When I try to lint/fix a query like this, I get "templating/parsing errors found" error: 
```sql
WITH cte AS (
    SELECT *
    FROM source
    WHERE col1 = 0
    DISTRIBUTE BY col1
),

SELECT *
FROM cte
```
I don't get errors without the WHERE condition or DISTRIBUTE BY

### Expected Behaviour

Sqlfluff should be able to lint/fix this query without errors.

### Observed Behaviour

Templating/parsing errors found.

### How to reproduce

Parse the query
```sql
WITH cte AS (
    SELECT *
    FROM source
    WHERE col1 = 0
    DISTRIBUTE BY col1
),

SELECT *
FROM cte
```

### Dialect

SparkSQL

### Version

sqlfluff, version 1.0.0

### Configuration

```
[sqlfluff]
dialect = databricks
templater = dbt
sql_file_exts = .sql
rules = core
exclude_rules = L022

[sqlfluff:indentation]
indented_joins = false
indented_using_on = true
template_blocks_indent = true

[sqlfluff:templater]
unwrap_wrapped_queries = true

[sqlfluff:templater:dbt]
project_dir = analytics_dbt_databricks/
profiles_dir = analytics_dbt_databricks/profiles/
profile = analytics_dbt
target = dev

[sqlfluff:rules]
tab_space_size = 4
max_line_length = 140
indent_unit = space
comma_style = trailing

[sqlfluff:rules:L010]
capitalisation_policy = upper

[sqlfluff:rules:L030]
extended_capitalisation_policy = lower
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by R7L208**
### Brief summary of the change made

- fixes #3484
- updates terminator grammar for various clauses to include `DISTRIBUTE BY`, `CLUSTER BY`, and `SORT BY`

### Are there any other side effects of this change that we should be aware of?
- Update terminator in `HAVING` for Snowflake, Ansi, and Hive
  - add terminator grammar that is defined in each dialect once
- Update terminator in `GROUP BY` for Snowflake, Ansi, and Hive
  - add terminator grammar that is defined in each dialect once

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

**Review by tunetheweb — CHANGES_REQUESTED** [review]
Change looks could but could I ask for one addition? It may turn out to be a bit of effort so understand if you'd prefer to merge as is.

File: src/sqlfluff/dialects/dialect_sparksql.py [hunk 1]
```diff
@@ -540,6 +559,20 @@
...
+                "DISTRIBUTE",
+                "CLUSTER",
+                "SORT",
+            ),
+            "BY",
+        ),
+        "LIMIT",
+        "HAVING",
+        "WINDOW",
+    ),
```
[the remark this answers is not in the document]
Comment (tunetheweb) [reply src/sqlfluff/dialects/dialect_sparksql.py:575]: Well Group By will now with this PR :-)

Agree Having doesn't (yet?) so can either leave that, or clean up for consistency.

[the remark this answers is not in the document]
Comment (R7L208) [reply src/sqlfluff/dialects/dialect_sparksql.py:575]: I'll clean up Having as well. It's implemented in a few other dialects similar by overriding the entire class

:)

**Review by tunetheweb — APPROVED** [review]
So much code clean up - I love it!!

**Pull request merged**